llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp

```diff
@@ -452,6 +456,13 @@ void NVPTXAsmPrinter::emitKernelFunctionDirectives(const Function &F,
                "should be 0 as well");
       }
     }
+
+    if (BlocksAreClusters && STI->getPTXVersion() >= 90) {
+      assert(!(ReqNTID.empty() || getClusterDim(F).empty()) &&
+             "blocksareclusters requires reqntid and cluster_dim");
+      O << ".blocksareclusters\n";
+    }
```

Comment:
Lets emit diagnostics here instead of asserting or silently omitting the annotation if the PTXVersion is too low. Also you can re-use the `ClusterDim` from above.

Reply:
A reasonable assumption to be sure, but I think since this error can occur due to malformed input we should not rely on asserts. 

Nit: use a DiagnosticInfoUnsupported so that you can attach the function. 

https://github.com/llvm/llvm-project/blob/2912c9c249350bae3178ba7fad605e1f269c32e7/llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp#L1822-L1826

Reply:
It looks like DiagnosticInfoUnsupported doesn't require that you provide a debug location and at least that way you can provide the function which is problematic.